compiler/src/iree/compiler/DispatchCreation/SetSplitReductionSizes.cpp

```diff
@@ -0,0 +1,121 @@
+  if (!linalgOp) {
+    return std::nullopt;
+  }
+  SmallVector<int64_t> dimSizes;
+  for (auto [loopRange, loopType] : llvm::zip_equal(
+           linalgOp.getStaticLoopRanges(), op.getLoopIteratorTypes())) {
+    if (loopType == utils::IteratorType::reduction) {
+      dimSizes.push_back(loopRange);
+    }
+  }
+  return dimSizes;
+}
+
+struct SetSplitReductionSizesPass final
+    : public impl::SetSplitReductionSizesPassBase<SetSplitReductionSizesPass> {
+  using Base::Base;
+  void runOnOperation() override {
+    getOperation()->walk([&](PartialReductionOpInterface tilingOp) {
+      std::optional<SmallVector<int64_t>> tileSizes =
+          getUserSpecifiedTileSize(tilingOp);
+      if (!tileSizes) {
+        return;
+      }
+      IREE::LinalgExt::setSplitReductionAttribute(tilingOp, tileSizes.value());
+    });
+  }
+
+private:
+  std::optional<SmallVector<int64_t>>
+  getUserSpecifiedTileSize(PartialReductionOpInterface op) const {
+    {
+      // First preference given to attribute set on the op.
+      std::optional<SmallVector<int64_t>> attributeTileSize =
+          IREE::LinalgExt::getSplitReductionSizes(op);
+      if (attributeTileSize) {
+        return attributeTileSize.value();
+      }
+    }
+
+    // Skip ops that aren't reductions.
+    unsigned numReduction = llvm::count_if(
+        op.getLoopIteratorTypes(), [](utils::IteratorType iteratorType) {
+          return iteratorType == utils::IteratorType::reduction;
+        });
+    if (numReduction == 0) {
+      return std::nullopt;
+    }
+
+    if (splitReductionTargetSize <= 0) {
+      return std::nullopt;
+    }
+    std::optional<SmallVector<int64_t>> opReductionSizes =
+        getStaticReductionDimSizes(op);
+    if (!opReductionSizes.has_value()) {
+      return std::nullopt;
+    }
+    auto findSmallestFactorWithLowerBound =
+        [](int64_t x, int64_t lowerBound) -> std::optional<int64_t> {
+      assert(x > 0);
+      assert(lowerBound > 0);
+      // We expect all numbers here to be relatively small, so just do trial
+      // division (with a limit just to be safe).
+      static constexpr int64_t kMaxIterations = 1 << 15;
+      int64_t upperBound = std::min(x, kMaxIterations);
+      for (int64_t i = lowerBound; i <= upperBound; i++) {
+        if (x % i == 0) {
+          return i;
+        }
+      }
+      return std::nullopt;
+    };
+
+    int64_t currentSplitReductionSize = 1;
+    SmallVector<int64_t> tileSizes(opReductionSizes->size());
+    // Tile dimensions until we reach or exceed the target. Tile sizes must
+    // divide the dimension size evenly, and we start with inner dimensions as
+    // we prefer tiling those.
+    for (int64_t i = tileSizes.size() - 1; i >= 0; i--) {
+      int64_t remainingSize =
+          llvm::divideCeil(splitReductionTargetSize, currentSplitReductionSize);
+      int64_t dimSize = (*opReductionSizes)[i];
+      if (dimSize == ShapedType::kDynamic) {
+        return std::nullopt;
+      }
```

Comment:
Nit: It would be useful if you could add some debug message for those nullopt return cases. It would help for debugging purpose.